### PR TITLE
[rom_e2e] Add rom_e2e_shutdown_alert_config

### LINKED ISSUE #14275: [rom-e2e] rom_e2e_shutdown_alert_config
**Testpoint name:** [rom_e2e_shutdown_alert_config](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson?q=rom_e2e_shutdown_alert_config)

**Contact person:** @alphan

**Description:** Verify that alerts trigger a chip reset when enabled.

- For PROD, PROD_END, DEV, and RMA life cycle states
  - Use an OTP image with an alert enabled.
- Boot a ROM_EXT that triggers this alert by writing to a ALERT_TEST register.
- Verify that ROM_EXT boots and the chip resets after the write to the ALERT_TEST
  register.

### PR DESCRIPTION
**Pull request by jon-flatley**
Adds the rom_e2e_shutdown_alert_config test.

Fixes #14275

### CODE REVIEW

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 1]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+                # Set to 0x739 to use the ROM_EXT hash measurement as the key
+                # manager attestation binding value.
+                "OWNER_SW_CFG_ROM_KEYMGR_ROM_EXT_MEAS_EN": "0x0",
+                # Report errors without any redaction.
+                "OWNER_SW_CFG_ROM_ERROR_REPORTING": "0xe2290aa5",
+                # Set the enables to kAlertEnableEnabled.
+                # See `alert_enable_t` in
+                # sw/device/silicon_creator/lib/drivers/alert.h
+                "OWNER_SW_CFG_ROM_ALERT_CLASS_EN": "0x07070707",
+                #"OWNER_SW_CFG_ROM_ALERT_CLASS_EN": "0xa9a9a9a9",
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:989]:
```suggestion
                # Enable class A alerts.
                # See `alert_enable_t` in
                # sw/device/silicon_creator/lib/drivers/alert.h
                "OWNER_SW_CFG_ROM_ALERT_CLASS_EN": "0x07a9a9a9",
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 2]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+                # Set the esclation policies to kAlertEscalateNone.
+                # See `alert_escalate_t` in
+                # sw/device/silicon_creator/lib/drivers/alert.h
+                "OWNER_SW_CFG_ROM_ALERT_ESCALATION": "0x76767676",
+                #"OWNER_SW_CFG_ROM_ALERT_ESCALATION": "0xd1d1d1d1",
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:994]:
```suggestion
                # Configure class A to escalate until phase 3 and disable escalation
                # for the remaining classes.
                # See `alert_escalate_t` in
                # sw/device/silicon_creator/lib/drivers/alert.h
                "OWNER_SW_CFG_ROM_ALERT_ESCALATION": "0x76d1d1d1",
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 3]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+                # Set the classifiactions to kAlertClassX.
+                # See `alert_class_t` in
+                # sw/device/silicon_creator/lib/drivers/alert.h
+                #"OWNER_SW_CFG_ROM_ALERT_CLASSIFICATION": ["0x94949494"] * 80,
+                "OWNER_SW_CFG_ROM_ALERT_CLASSIFICATION": ["0xeeeeeeee"] * 80,
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:999]:
```suggestion
                # Classify UART0 alerts (alert source 0) as class A and leave others as unconfigured.
                # See `alert_class_t` in
                # sw/device/silicon_creator/lib/drivers/alert.h
                "OWNER_SW_CFG_ROM_ALERT_CLASSIFICATION": ["0xee949494"] + ["0x94949494"] * 79,
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 4]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+                # Set the classifiactions to kAlertClassX. See `alert_class_t` in
+                # sw/device/silicon_creator/lib/drivers/alert.h
+                #"OWNER_SW_CFG_ROM_LOCAL_ALERT_CLASSIFICATION": ["0x94949494"] * 16,
+                "OWNER_SW_CFG_ROM_LOCAL_ALERT_CLASSIFICATION": ["0x94949494"] * 16,
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1003]:
```suggestion
                # Leave local alert classification as unconfigured. See `alert_class_t` in
                # sw/device/silicon_creator/lib/drivers/alert.h
                "OWNER_SW_CFG_ROM_LOCAL_ALERT_CLASSIFICATION": ["0x94949494"] * 16,
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 5]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+                    "0xFFFFFFFF",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                    "0x0",
+                ],
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1027]:
```suggestion
                # Set the alert phase cycles to 0 for class A and to 0xffffffff for other classes.
                "OWNER_SW_CFG_ROM_ALERT_PHASE_CYCLES": ["0x0"] * 4  + ["0xffffffff"] * 28,
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 6]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+            },
+        ),
+    ],
+)
+
+otp_alert_digest(
+    name = "shutdown_alert_digest_cfg",
+    otp_img = ":shutdown_alert_owner_sw_cfg",
+)
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1036]: Nice!

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 7]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+        ":shutdown_alert_digest_cfg",
+    ],
+) for lc_state in ALERT_LC_STATES]
+
+[bitstream_splice(
+    name = "bitstream_shutdown_alert_{}".format(lc_state.lower()),
+    src = "//hw/bitstream:rom",
+    data = ":otp_img_shutdown_alert_{}".format(lc_state.lower()),
+    meminfo = "//hw/bitstream:otp_mmi",
+    tags = ["vivado"] + maybe_skip_in_ci(lc_state_val),
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1052]: Thanks for using `maybe_skip_in_ci()`.

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 8]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+) for lc_state in ALERT_LC_STATES]
+
+[
+    opentitan_functest(
+        name = "shutdown_alert_{}".format(
+            lc_state.lower(),
+        ),
+        cw310 = cw310_params(
+            bitstream = ":bitstream_shutdown_alert_{}".format(lc_state.lower()),
+            rom_kind = "testrom",
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1080]: I think this is a typo. We want to run this with the actual rom, which is already spliced into the bitstream.

```suggestion
```

File: sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c [hunk 9]
```diff
@@ -0,0 +1,46 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c:9]:
```suggestion
#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
```

File: sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c:6]:
```suggestion
```

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 11]
```diff
@@ -967,6 +967,138 @@ test_suite(
...
+        "fpga_cw310",
+    ],
+    deps = [
+        "//hw/ip/uart/data:uart_regs",
+        "//sw/device/lib/base:abs_mmio",
+        "//sw/device/lib/base:bitfield",
+        "//sw/device/lib/testing/test_framework:ottf_ld_silicon_creator_slot_a",
+        "//sw/device/lib/testing/test_framework:ottf_main",
+        "//sw/device/silicon_creator/lib/drivers:retention_sram",
+        "//sw/device/silicon_creator/lib/drivers:rstmgr",
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/BUILD:1069]:
```suggestion
        "//hw/ip/uart/data:uart_regs",
        "//sw/device/lib/base:abs_mmio",
        "//sw/device/lib/testing/test_framework:ottf_ld_silicon_creator_slot_a",
        "//sw/device/lib/testing/test_framework:ottf_main",
        "//sw/device/silicon_creator/lib/drivers:rstmgr",
```

**Review by alphan — APPROVED** [review]
Thanks @jon-flatley! This looks pretty good to me, I just recommend tightening the alert config so that it only resets when it receives a signal from UART -- just to reduce the chances of false positives. The rest are minor nits.

File: sw/device/silicon_creator/rom/e2e/BUILD (hunk 11 above)
Comment (jon-flatley) [reply sw/device/silicon_creator/rom/e2e/BUILD:1069]: I think we want the retention_sram driver here so we can query the reset reason

File: sw/device/silicon_creator/rom/e2e/BUILD (hunk 5 above)
Comment (jon-flatley) [reply sw/device/silicon_creator/rom/e2e/BUILD:1027]: Should this be `["0x0"] * 4  + ["0xffffffff"] * 12,`? It's my understanding that there are `PARAM_N_PHASES * PARAM_N_CLASSES = 4 * 4 = 16` fields here.

File: sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c (hunk 9 above)
Comment (alphan) [reply sw/device/silicon_creator/rom/e2e/rom_e2e_shutdown_alert_config_test.c:8]: nevermind

File: sw/device/silicon_creator/rom/e2e/BUILD (hunk 5 above)
Comment (alphan) [reply sw/device/silicon_creator/rom/e2e/BUILD]: Yeah, I think you're right.

File: sw/device/silicon_creator/rom/e2e/BUILD (hunk 1 above)
Comment (alphan) [reply sw/device/silicon_creator/rom/e2e/BUILD]: same here, wrong endianness

**Comment by alphan** [comment]
@jon-flatley I rebased, squashed your commits (original + update, preserved your authorship of course), and added a commit that adds the alert constants to `const.bzl` and fixes the values in the OTP configuration. All tests pass on my machine.

```
//sw/device/silicon_creator/rom/e2e:shutdown_alert_dev_fpga_cw310_rom (cached) PASSED in 19.1s
//sw/device/silicon_creator/rom/e2e:shutdown_alert_prod_end_fpga_cw310_rom (cached) PASSED in 19.1s
//sw/device/silicon_creator/rom/e2e:shutdown_alert_prod_fpga_cw310_rom (cached) PASSED in 19.1s
//sw/device/silicon_creator/rom/e2e:shutdown_alert_rma_fpga_cw310_rom (cached) PASSED in 19.1s
```

Marking as ready.

**Pull request merged**